### PR TITLE
`@blockprotocol/hook` package

### PR DESCRIPTION
**Pull request by nathggns**
## 🌟 What is the purpose of this PR?

This adds a `@blockprotocol/hook` package which has some helpful utilities for a block/embedding application to implement the hook service as described at https://github.com/blockprotocol/blockprotocol/blob/main/rfcs/text/0430-hook-service.md

**Please ignore the incorrect branch name.**

## 🔗 Related links
- [Github discussion](https://github.com/blockprotocol/blockprotocol/discussions/475)
- [Asana task](https://app.asana.com/0/1201095311341924/1202698219459270/f) (_internal_)
- [Notion doc](https://www.notion.so/hashintel/Hook-service-spec-1fcc185dc67c4ce7ace06878a144977d) (_internal_)

## ❓ How to test this?

This is in use in https://github.com/hashintel/hash/tree/nh/hook-service-implementation – take a look at that.

There are no automated tests.

### CODE REVIEW

File: packages/@blockprotocol/hook/src/react.ts [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+): { hookService: HookBlockHandler | null } => {
+  return useHookServiceConstructor({
+    Handler: HookBlockHandler,
+    constructorArgs,
+    ref,
+  });
+};
+
+/**
+ * Create a HookBlockHandler instance, using a reference to an element in the
```
Comment (CiaranMn) [inline packages/@blockprotocol/hook/src/react.ts:75]: This was originally my error in the graph service - 
```suggestion
 * Create a HookEmbedderHandler instance, using a reference to an element around the
```

File: packages/@blockprotocol/hook/src/react.ts [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+  teardown: (() => Promise<void>) | null;
+  params: {
+    service: HookBlockHandler | null;
+    node: T;
+    type: string;
+    path: string;
+  };
+};
+
+export const useHook = <T extends HTMLElement>(
```
Comment (CiaranMn) [inline packages/@blockprotocol/hook/src/react.ts:107]: Can you add a docstring on this? I'm keen that absolutely every public-facing function or method in these packages has some useful comment on it.

File: packages/@blockprotocol/hook/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+For React, we provide a `useHookBlockService` hook, which accepts a `ref` to an element. This will return an object with the shape of `{ hookService: HookBlockHandler | null }` which you can use to send hook messages.
+
+We also provide a `useHook` hook to make sending hook messages easier.
+
+```typescript
+import { useHook } from "@blockprotocol/hook/react";
+
+useHook(hookService, nodeRef, "text", "$.text", (node) => {
+  node.innerText = "hook fallback";
```
Comment (CiaranMn) [inline packages/@blockprotocol/hook/README.md:41]: It'd be helpful to add a comment here explaining that this function will be run if the EA doesn't implement the hook service, or the particular type of hook requested. 

Or mention that above somewhere.

File: packages/@blockprotocol/hook/src/types.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+export type HookResponse = {
+  hookId: string;
+};
+
+export type HookData = {
+  node: HTMLElement | null;
+  type: string;
+  path: string;
+  hookId?: string | null;
```
Comment (CiaranMn) [inline packages/@blockprotocol/hook/src/types.ts:11]: The JSON schema and the RFC said this should be `null` on first call, not omitted or null. We can change that but the JSON schema will need to be changed.
```suggestion
  hookId: string | null;
```

File: packages/@blockprotocol/hook/src/react.ts [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+            try {
+              hook.id = null;
+              if (hookRef.current === hook) {
+                hookRef.current = null;
+              }
+
+              if (!service.destroyed) {
+                await service.hook({
+                  data: {
+                    hookId: hook.id,
```
Comment (CiaranMn) [inline packages/@blockprotocol/hook/src/react.ts:180]: This was set `null` a few lines above

**Review by CiaranMn — COMMENTED** [review]
Our first new service! I love to see it

A few small comments on the code – I will test tomorrow by looking at the counterpart linked PR

File: packages/@blockprotocol/hook/src/react.ts [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+};
+
+export const useHook = <T extends HTMLElement>(
+  service: HookBlockHandler | null,
+  ref: RefObject<T | null | void>,
+  type: string,
+  path: string,
+  fallback: (node: T) => void | (() => void),
+) => {
+  const hookRef = useRef<null | Hook<T>>(null);
```
Comment (CiaranMn) [inline packages/@blockprotocol/hook/src/react.ts:114]: Maybe `existingHook`?

File: packages/@blockprotocol/hook/README.md (hunk 3 above)
Comment (nathggns) [reply packages/@blockprotocol/hook/README.md:41]: https://github.com/blockprotocol/blockprotocol/commit/931e03c0ee74cf4606cc3df0f6d40d57d5c554fc

File: packages/@blockprotocol/hook/src/react.ts (hunk 2 above)
Comment (nathggns) [reply packages/@blockprotocol/hook/src/react.ts:121]: https://github.com/blockprotocol/blockprotocol/commit/931e03c0ee74cf4606cc3df0f6d40d57d5c554fc

File: packages/@blockprotocol/hook/src/react.ts (hunk 5 above)
Comment (nathggns) [reply packages/@blockprotocol/hook/src/react.ts]: Good catch https://github.com/blockprotocol/blockprotocol/commit/567ed9c2fe9908a5f5447019c1bc3480f2c57258

File: packages/@blockprotocol/hook/src/types.ts (hunk 4 above)
Comment (nathggns) [reply packages/@blockprotocol/hook/src/types.ts:11]: I think it should be included always. Good catch.

File: packages/@blockprotocol/hook/src/react.ts (hunk 6 above)
Comment (nathggns) [reply packages/@blockprotocol/hook/src/react.ts]: https://github.com/blockprotocol/blockprotocol/commit/60a35b59f5ee55bd6c01dc138aa7230e46a245ce

**Review by CiaranMn — APPROVED** [review]
Welcome, hook service! 🥳

**Pull request merged**